nifi-commons/nifi-schema-utils/src/main/java/org/apache/nifi/repository/schema/RecordSchema.java

```diff
@@ -0,0 +1,188 @@
+                final RecordField field = readField(dis);
+                fields.add(field);
+            }
+
+            return new RecordSchema(fields);
+        } catch (final IOException ioe) {
+            throw new IOException("Unable to read Record Schema from stream", ioe);
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static RecordField readField(final DataInputStream dis) throws IOException {
+        final Map<String, Object> schemaFieldMap = new HashMap<>();
+        final int numElementsToRead = dis.readInt();
+        for (int i = 0; i < numElementsToRead; i++) {
+            final String fieldName = dis.readUTF();
+            final String typeName = dis.readUTF();
+            Object fieldValue = null;
+
+            switch (typeName) {
+                case STRING_TYPE:
+                    fieldValue = dis.readUTF();
+                    break;
+                case INT_TYPE:
+                    fieldValue = dis.readInt();
+                    break;
+                case LONG_TYPE:
+                    fieldValue = dis.readLong();
+                    break;
+                case SUBFIELD_TYPE: {
+                    final int numFields = dis.readInt();
+                    final List<RecordField> subFields = new ArrayList<>(numFields);
+                    for (int j = 0; j < numFields; j++) {
+                        subFields.add(readField(dis));
+                    }
+                    fieldValue = subFields;
+                    break;
+                }
+                default: {
+                    throw new IOException("Cannot read schema because the schema definition contains a field named '"
+                        + fieldName + "' with a Field Type of '" + typeName + "', which is not a known Field Type");
+                }
+            }
+
+            schemaFieldMap.put(fieldName, fieldValue);
+        }
+
+        final String fieldName = (String) schemaFieldMap.get(FIELD_NAME);
+        final String fieldTypeName = (String) schemaFieldMap.get(FIELD_TYPE);
+        final String repetitionName = (String) schemaFieldMap.get(REPETITION);
+        List<RecordField> subFields = (List<RecordField>) schemaFieldMap.get(SUBFIELDS);
+        if (subFields == null) {
+            subFields = Collections.emptyList();
+        }
+
+        final Repetition repetition = Repetition.valueOf(repetitionName);
+        if (FieldType.COMPLEX.name().equals(fieldTypeName)) {
+            return new ComplexRecordField(fieldName, repetition, subFields);
+        } else if (FieldType.UNION.name().equals(fieldTypeName)) {
+            return new UnionRecordField(fieldName, repetition, subFields);
+        } else if (FieldType.MAP.name().equals(fieldTypeName)) {
+            if (subFields.size() != 2) {
+                throw new IOException("Found a Map that did not have a 'Key' field and a 'Value' field but instead had " + subFields.size() + " fields: " + subFields);
```

Comment:
IMHO message is a bit confusing. I'd say "Field of type Map must only contain 2 elements representing key/value. Was. . . ."